### ISSUE TITLE
Python Formating and Guidelines

### ISSUE THREAD

**Issue by JoaoRodrigues** [openmm/openmm#2449]
Would it be interesting or advantageous in your opinion to clean up a bit the Python code to make it a bit more readable? Some 'issues' I've found when reading/editing the code in a terminal windows for instance are:

* incredibly long lines, some over 150 chars long.
* lack of space around operators: e.g. `if residue.name!='HOH' and residue.name!='WAT':`
* unused variables: e.g. `invBox` on line 177 in modeller.py
* `from ... import *` imports
* unused imports: e.g. simtk.unit.gram in modeller.py

Biopython recently ran all their code through [black](https://black.readthedocs.io/en/stable/) and while I am not very fond of automatic anything tools, it required very little correction afterwards. Would you be interested in doing something like this, over time?

**Comment by jchodera** [comment]
Great idea, but I think the might be the least of our worries about the Python layer:
* Depends on a fragile custom doxygen-dependent script to generate swig wrappers
* Not PEP8
* Utterly un-Pythonic API

I think it would be useful to think about how we might reconceptualize the Python layer for a future major release

**Comment by JoaoRodrigues** [comment]
Honestly, I don't see it as very unPythonic. It's quite straightforward and very explicit. Compared to other APIs out there in the comp bio/chem world, it's actually really nice! Not to say it couldn't use some improvements, but what exists works and works well so.. 

It's not PEP8, clearly, that's what I was suggesting to fix. It's a more achievable goal in the near future and would help a lot with readability. Same for checking test coverage for the Python API and work on increasing it.

Just a suggestion!

**Comment by jchodera** [comment]
It's "un-Pythonic" in the sense that it uses idioms borrowed from C++, rather than Python. For example, iterating over particles looks like:
```python
for index in range(system.getNumParticles()):
    mass = system.getParticleMass(index)
    # do something...
```
rather than
```python
for  mass in system.particle_masses:
    # do something...
```

`CamelCase` is also a distinctly C++ style feature; PEP8 specifies [`snake_case`](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables) should be used for method and variable names instead. 

A cleanup of the current Python API code would have to then be a weird mix of PEP8 and C++ to avoid breaking anything, so I'm not sure it's worth it at this point. Instead, investment in the design for a new Pythonic PEP8-compliant Python layer alternative---perhaps using the more modern and maintainable [pybind11](https://pybind11.readthedocs.io/en/stable/)---might be a better investment.

**Comment by peastman** [comment]
I'm definitely against trying to make it strictly pep8 compliant.  Consistency between the C++ and Python APIs is much more important in my mind.  Besides, camel case is very common among Python APIs, including a bunch of packages in the standard library.  They've chosen to leave them like that because they consider backward compatibility to be more important.

I am ok with the idea of adding methods to the Python wrappers to allow more idiomatic code.  For example, the Python version of System has a `getForces()` method that doesn't exist in the C++ version.  We could consider adding similar methods in other places.

But to get back to @JoaoRodrigues original proposal, I think that's a fine idea.  It's just a cleanup to internal code, so it doesn't affect compatibility.  I'd prefer to avoid automated code formatters.  The DeepChem project runs all its code through yapf, and it has a strong tendency to take clean, readable code and do horrible things with it, just because the original code didn't match its arbitrary rules.  There are times when a single very long line actually is the best way to format something!  But if done judiciously, some code cleanup would be a good thing.

**Comment by JoaoRodrigues** [comment]
I agree with you. In an ideal world, we would have a fully-PEP8 compliant API. 

However, as it is, I am willing to compromise on some things (e.g. camelCase) to be able to access super high-performance C++ code. At the end of the day, open-source projects with small developer communities have to compromise on what to spend their energy on.  

I was more talking about the 'pure' Python modules, such as modeller.py or PDBFile.py, that are mostly independent from the C++ code (except for function calls).

**Comment by tristanic** [comment]
@jchodera a +1 for me on the PyBind11 suggestion! Having been through the process of wrapping a large C++ library in SWIG then re-doing the whole thing in PyBind11 (https://github.com/tristanic/chimerax-clipper), in my opinion the difference in ease-of-use is like night and day. Not just in terms of clarity of code (the weird amalgam of C++, Python and SWIG syntax in a single file always gave me enormous headaches), but perhaps equally importantly in ease of building - header only, all done through the compiler with no external packages needed.
